### PR TITLE
Add wildcard and fuzzy query translation rules.

### PR DESCRIPTION
**Pull request by Shivani-techno**
### Description
Adds wildcard and fuzzy query translation support to the Solr-to-OpenSearch query engine.

Previously, `fieldRule` threw errors when encountering wildcard (`*`, `?`) or fuzzy (`~`, `~N`) patterns in field:value queries. This PR replaces those errors with proper OpenSearch query DSL translation:

- `title:te?t` → `{"wildcard": {"title": {"value": "te?t"}}}`
- `title:test*` → `{"wildcard": {"title": {"value": "test*"}}}`
- `title:roam~` → `{"fuzzy": {"title": {"value": "roam"}}}`
- `title:roam~2` → `{"fuzzy": {"title": {"value": "roam", "fuzziness": 2}}}`

No parser or AST changes needed — wildcards and fuzzy suffixes are already parsed as part of `FieldNode.value` by the PEG grammar. The change is entirely within `fieldRule.ts`.

### Issues Resolved

Resolves the TODO items in `fieldRule.ts`:
- "Add support for wildcard queries"
- "Add support for fuzzy queries"

### Testing

- 21 unit tests in `fieldRule.test.ts` (5 wildcard, 4 fuzzy, 3 non-fuzzy tilde edge cases, 2 boost compatibility, plus existing match/exists/leaf tests)
- All 483 tests in the SolrTransformations test suite pass
- Output uses expanded Map form compatible with `boostRule` for boost injection

### Check List
- [ ] New functionality includes testing
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/query-engine/transformer/rules/fieldRule.ts [hunk 1]
```diff
@@ -1,69 +1,66 @@
...
-    console.error(msg);
-    throw new Error(msg);
+  // Fuzzy search: field:roam~ or field:roam~2 → fuzzy query
+  // Check fuzzy before wildcard since ~ is more specific
+  const fuzzyMatch = FUZZY_PATTERN.exec(value);
+  if (fuzzyMatch) {
+    const term = fuzzyMatch[1];
+    const distance = fuzzyMatch[2];
+    const fuzzyParams = new Map<string, any>([['value', term]]);
+    if (distance) {
```
Comment (niravpi) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/query-engine/transformer/rules/fieldRule.ts]: Nit: `if (distance !== '')` reads better than `if (distance)`. It works today because `"0"` is truthy in JS, but the intent ("was a digit present?") is clearer with an explicit empty-string check.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/query-engine/transformer/rules/fieldRule.ts [hunk 2]
```diff
@@ -1,69 +1,66 @@
...
+      fuzzyParams.set('fuzziness', parseInt(distance, 10));
```
Comment (niravpi) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/query-engine/transformer/rules/fieldRule.ts]: AFAIU from the docs, OpenSearch only accepts fuzziness 0, 1, 2, or `"AUTO"`. A query like `roam~9` would produce `{"fuzziness": 9}` which OpenSearch rejects at query time. Should we consider either validating here (Solr also only allows 0-2) or adding a comment that we intentionally let OpenSearch handle the rejection.

**Comment by niravpi** [comment]
Unit tests cover `fieldRule` in isolation, but there's no test verifying the full round-trip: PEG parser → AST → transformer → DSL. The parser producing correct `FieldNode.value` for wildcard/fuzzy patterns is assumed but unverified. #2688 established the e2e framework in `cases.testcase.ts` adding a few positive cases (`title:te?t`, `title:test*`, `title:roam~`, `title:roam~2`) would close this gap and verify against real Solr.

**Comment by niravpi** [comment]
#2688's sandbox results categorize "Wildcard/fuzzy/proximity (6)" as expected rejections (part of the 79 rejection count). Once this PR merges, those queries will start succeeding, can you please try to run the sandbox validation once and see if Wildcard/fuzzy query succeed ?

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/query-engine/transformer/rules/fieldRule.ts (hunk 1 above)
Comment (Shivani-techno) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/query-engine/transformer/rules/fieldRule.ts]: Done,  updated to if (distance !== '').

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/query-engine/transformer/rules/fieldRule.ts (hunk 2 above)
Comment (Shivani-techno) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/query-engine/transformer/rules/fieldRule.ts]: Done, added validation to clamp fuzziness to max 2 using Math.min(fuzziness, 2), matching both Solr and OpenSearch limits. Also added a test case for roam~9 verifying it produces fuzziness: 2.

**Comment by niravpi** [comment]
Looks like integration test comment is pending can you please respond back to the same?
Also did we tried validating via sandbox validation ?

**Comment by niravpi** [comment]
Based on the code seems like the fuzzy logic in `fieldRule` will not be used as  `~` is not in the PEG grammar's `valueChars` (`$[a-zA-Z0-9._*#$@?/]+`), so `title:roam~2` fails at the parser before `fieldRule` ever sees it. Unit tests pass because they construct `FieldNode` manually, bypassing the parser.
Can you please help to confirm this behaviour once ?

**Comment by Shivani-techno** [comment]
> Unit tests cover `fieldRule` in isolation, but there's no test verifying the full round-trip: PEG parser → AST → transformer → DSL. The parser producing correct `FieldNode.value` for wildcard/fuzzy patterns is assumed but unverified. [#2688](https://github.com/opensearch-project/opensearch-migrations/pull/2688) established the e2e framework in `cases.testcase.ts` adding a few positive cases (`title:te?t`, `title:test*`, `title:roam~`, `title:roam~2`) would close this gap and verify against real Solr.

Added 4 e2e test cases in wildcard-fuzzy-query.testcase.ts following the framework from #2688: title:te?t, title:search*, title:roam~, and title:roam~2. These verify the full PEG parser → AST → transformer → DSL round-trip. Also found and fixed a gap in the PEG grammar — ~ was missing from valueChars, which caused fuzzy patterns to fail at parse time.

**Comment by Shivani-techno** [comment]
> [#2688](https://github.com/opensearch-project/opensearch-migrations/pull/2688)'s sandbox results categorize "Wildcard/fuzzy/proximity (6)" as expected rejections (part of the 79 rejection count). Once this PR merges, those queries will start succeeding, can you please try to run the sandbox validation once and see if Wildcard/fuzzy query succeed ?

Ran sandbox validation — wildcard and fuzzy queries now succeed through the shim.

**Comment by Shivani-techno** [comment]
> Looks like integration test comment is pending can you please respond back to the same? Also did we tried validating via sandbox validation ?

Yes, both done.

**Comment by Shivani-techno** [comment]
> Based on the code seems like the fuzzy logic in `fieldRule` will not be used as `~` is not in the PEG grammar's `valueChars` (`$[a-zA-Z0-9._*#$@?/]+`), so `title:roam~2` fails at the parser before `fieldRule` ever sees it. Unit tests pass because they construct `FieldNode` manually, bypassing the parser. Can you please help to confirm this behaviour once ?

Fixed in the latest commit by adding ~ to the grammar's valueChars rule: $[a-zA-Z0-9._*#$@?/~]+. Confirmed working end-to-end via sandbox — fuzzy queries now parse correctly and produce the expected DSL.

**Comment by Shivani-techno** [comment]
> > Based on the code seems like the fuzzy logic in `fieldRule` will not be used as `~` is not in the PEG grammar's `valueChars` (`$[a-zA-Z0-9._*#$@?/]+`), so `title:roam~2` fails at the parser before `fieldRule` ever sees it. Unit tests pass because they construct `FieldNode` manually, bypassing the parser. Can you please help to confirm this behaviour once ?

Fixed in the latest commit by adding ~ to the grammar's valueChars rule: `$[a-zA-Z0-9._*#$@?/~]+`. Confirmed working end-to-end via sandbox — fuzzy queries now parse correctly and produce the expected DSL.

The double backticks (``) around the regex prevent GitHub from interpreting # and $ as special characters.

**Review by niravpi — APPROVED** [review]
